lightning/src/chain/channelmonitor.rs

```diff
@@ -2039,11 +2205,28 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 
 			self.is_paying_spendable_output(&tx, height, &logger);
 		}
+
+		self.block_confirmed(height, txn_matched, watch_outputs, claimable_outpoints, broadcaster, fee_estimator, logger)
```

Comment:
Isn't this redundant with `block_confirmed` within `update_best_block` or does the API assume non-block-connected clients can't call `update_best_block` ?

I think that's okay as a temporary location because you need `txn_matched` for `OnchainTxHandler::update_claims_view` and we might need to split `update_claims_view` in two halves `update_best_block`/`event_confirmed`.

I can do it as a follow-up if you wanna :)

Reply:
> I suppose it may be redundant. @TheBlueMatt If a user must call transaction_confirmed for relevant transactions, why does update_best_block need to call block_confirmed? I think we discussed this earlier in an earlier conversation.

Hmm, your link is dead, so its entirely possible I said something different, but I think it doesnt. I guess really its a question of what the API should be, on the `ChannelManager` side we kinda went with (but can change) "you call them in any order, but you always have to call both transactions_confirmed and update_best_block". We could drop the update_best_block call if you've called transactions_confirmed, making this line required, but I think we'll have users who want to call both in many cases, so we can't/shouldn't move to "you cannot call both, only one".

Reply:
For reference, I think this is what I had tried to link: https://github.com/rust-bitcoin/rust-lightning/pull/858#discussion_r604419283.

IIUC, the question here is not whether a user should call `update_best_block` but rather if `update_best_block` should call `block_confirmed` (i.e., to handle height-based triggers described in the linked comment).

Reply:
> but rather if update_best_block should call block_confirmed

Hmm? This line is in `transactions_confirmed`, though. I read the question as should we handle the height-based triggers in `transactions_confirmed`, even if the user doesn't call `update_best_block`.

Reply:
Ah, maybe I'm answering the question the other way around then. :) Either way, isn't it still redundant for it to be called in both places? The user needs to call `transactions_confirmed` for each transaction and also call `update_best_block` either at the height or at some later height.

> We could drop the update_best_block call if you've called transactions_confirmed, making this line required, but I think we'll have users who want to call both in many cases, so we can't/shouldn't move to "you cannot call both, only one".

Back to what you had said quoted above, I wasn't saying to change the API to say not call both but rather change how `update_best_block` is implemented (i.e., don't have `update_best_block` call `block_confirmed`).

Reply:
> Back to what you had said quoted above, I wasn't saying to change the API to say not call both but rather change how update_best_block is implemented (i.e., don't have update_best_block call block_confirmed).

Right, I think `update_best_block` needs to call `block_confirmed` because sometimes you *only* get `update_best_block` (ie if there are no relevant transactions confirmed). The converse isn't true (ie you do always get `update_best_block` even if you also get `transactions_confirmed`), though we could change that if we wanted to.

Reply:
Gotcha... there could be transactions maturing and we need to handle the associated events.

Reply:
Right, the question to me is do we want to support *only* calling `transactions_confirmed`. I don't really care either way, honestly.

Reply:
So in that world would you call `transactions_confirmed` with empty `txdata` in lieu of `update_best_block`? Doesn't that bring us back full circle to the equivalent of calling `block_connected`? Or did I misunderstand / am missing something about the reorg case?

Reply:
Hmm, I wasn't assuming going that far, I dunno, its hard. Thinking about it from the client's perspective, having two calls I think is useful - I believe it wouldn't be strange for an electrum client to receive both a notification that a new block is connected and a notification that transactions matched the installed filter, and ideally you'd be able to just map 1:1 notifications to LDK calls.

Reply:
Yeah I think here the question of how we refactor `block_confirmed` to extract only what matters in case of new transactions confirmed is dependent on the API we want to expose to an Electrum user. If we go for a 1:1 mapping, we can repeat the `onchain_tx_handler.update_claims_view()` in `transaction_confirmed` and don't care about the rest of the method. The remaining logic of `block_confirmed` is height-triggered so shouldn't be impacted. At _first sight_

Reply:
Agreed that a 1:1 mapping would be preferable.

Reply:
I tried this but many tests failed. Haven't looked into them though. Is this the change you had in mind?

```diff
diff --git a/lightning/src/chain/channelmonitor.rs b/lightning/src/chain/channelmonitor.rs
index 777bc06b..f73e7630 100644
--- a/lightning/src/chain/channelmonitor.rs
+++ b/lightning/src/chain/channelmonitor.rs
@@ -2205,7 +2205,29 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
                        self.is_paying_spendable_output(&tx, height, &logger);
                }
 
-               self.block_confirmed(height, txn_matched, watch_outputs, claimable_outpoints, broadcaster, fee_estimator, logger)
+               self.onchain_tx_handler.update_claims_view(&txn_matched, claimable_outpoints, Some(height), &&*broadcaster, &&*fee_estimator, &&*logger);
+
+               // Determine new outputs to watch by comparing against previously known outputs to watch,
+               // updating the latter in the process.
+               watch_outputs.retain(|&(ref txid, ref txouts)| {
+                       let idx_and_scripts = txouts.iter().map(|o| (o.0, o.1.script_pubkey.clone())).collect();
+                       self.outputs_to_watch.insert(txid.clone(), idx_and_scripts).is_none()
+               });
+               #[cfg(test)]
+               {
+                       // If we see a transaction for which we registered outputs previously,
+                       // make sure the registered scriptpubkey at the expected index match
+                       // the actual transaction output one. We failed this case before #653.
+                       for tx in &txn_matched {
+                               if let Some(outputs) = self.get_outputs_to_watch().get(&tx.txid()) {
+                                       for idx_and_script in outputs.iter() {
+                                               assert!((idx_and_script.0 as usize) < tx.output.len());
+                                               assert_eq!(tx.output[idx_and_script.0 as usize].script_pubkey, idx_and_script.1);
+                                       }
+                               }
+                       }
+               }
+               watch_outputs
        }
 
        fn block_confirmed<B: Deref, F: Deref, L: Deref>(
```